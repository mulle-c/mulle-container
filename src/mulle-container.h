//
//  mulle-container.h
//  mulle-container
//
//  Created by Nat! on 02/11/15.
//  Copyright (c) 2015 Nat! - Mulle kybernetiK.
//  Copyright (c) 2015 Codeon GmbH.
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  Neither the name of Mulle kybernetiK nor the names of its contributors
//  may be used to endorse or promote products derived from this software
//  without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
//  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
//  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
//  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
//  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
//  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
//  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
//  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//
#ifndef mulle_container__h__
#define mulle_container__h__

#define MULLE_CONTAINER_VERSION  ((2 << 20) | (0 << 8) | 0)

#include "include.h"

#include "mulle-container-operation.h"
#include "mulle-container-math.h"
#include "mulle-container-callback.h"
#include "mulle-container-callback-global.h"

#include "mulle-prime.h"
#include "mulle-hash.h"

#include "mulle--array.h"
#include "mulle-array.h"

#include "mulle--set.h"
#include "mulle-set.h"

#include "mulle--map.h"
#include "mulle-map.h"

#include "mulle-range.h"
#include "mulle--rangeset.h"

#include "mulle--pointerqueue.h"
#include "mulle-pointerarray.h"
#include "mulle--pointermap.h"
#include "mulle-pointerset.h"
#include "mulle-pointerpairarray.h"

#include "mulle-container-callback.h"
#include "mulle-container-operation.h"

#if MULLE_C11_VERSION < ((3 << 20) | (2 << 8) | 0)
# error "mulle-c11 is too old"
#endif

#if MULLE_ALLOCATOR_VERSION < ((2 << 20) | (0 << 8) | 0)
# error "mulle-allocator is too old"
#endif


#endif /* mulle_container_h */
