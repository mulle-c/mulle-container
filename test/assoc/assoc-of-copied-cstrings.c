//
//  main.c
//  test-assoc
//
//  Created by Nat! on 04.11.15.
//  Copyright (c) 2015 Nat! - Mulle kybernetiK.
//  Copyright (c) 2015 Codeon GmbH.
//  All rights reserved.
//

#include <mulle-container/mulle-container.h>
#include <mulle-testallocator/mulle-testallocator.h>

#include <stdio.h>


// MFing strings are const, though noone asked for that
static char   *test_strings[] =
{
   "Es haben viel fröhliche Menschen",
   "Lang vor uns gelebt",
   "und gelacht",
   "Den Ruhenden unter dem Rasen",
   "Sei fröhlich",

   "der Becher gebracht.",
   "Es werden viel fröhliche Menschen",
   "Lang nach uns des Lebens sich freun",
   "Uns Ruhenden unter dem Rasen",
   "Den Becher der Fröhlichkeit weihn."
};


static void  loops( void)
{
   struct mulle_assoc                        *assoc;
   struct mulle_assocenumerator              rover;
   struct mulle_pointerpair                  item;
   int                                       i;
   struct mulle_container_keyvaluecallback   callback;
   void                                      *key;
   void                                      *value;

   callback.keycallback    = mulle_container_keycallback_copied_cstring;
   callback.valuecallback  = mulle_container_valuecallback_copied_cstring;
   //
   // the string is "strdup" by the assoc on add, and will be freed when
   // removed or destroyed
   //
   assoc = mulle_assoc_create( 0,
                               &callback,
                               _mulle_pointerpair_compare_string_key,
                               NULL);

   for( i = 0; i < 10; i++)
      mulle_assoc_add( assoc, test_strings[ i], test_strings[ 10 - 1 - i]);

   for( i = 0; i < 10; i++)
      test_strings[ i][ 0] = 0; // destroy original

   printf( "Get (should be already sorted):\n");
   for( i = 0; i < 10; i++)
   {
      item = mulle_assoc_get_at_index( assoc, i);
      printf( "\t%s=%s\n", (char *) item.key, (char *) item.value);
   }
   printf( "\n");

   printf( "Enumeration:\n");
   rover = mulle_assoc_enumerate( assoc);
   while( mulle_assocenumerator_next( &rover, &key, &value))
      printf( "\t%s=%s\n", (char *) key, (char *) value);
   mulle_assocenumerator_done( &rover);
   printf( "\n");

   mulle_assoc_destroy( assoc);
}



// the mulle_testallocator detects and aborts on leaks
static void  run_test( void (*f)( void), char *name)
{
   mulle_testallocator_reset();
#ifdef DEBUG
   fprintf( stderr, "%s\n", name);
#endif
   (f)();
   mulle_testallocator_reset();
}


int main( int argc, const char * argv[])
{
   size_t  i;

   for( i = 0; i < 10; i++)
      test_strings[ i] = strdup( test_strings[ i]);

   run_test( loops, "loops");

   for( i = 0; i < 10; i++)
      free( test_strings[ i]);

   return( 0);
}
