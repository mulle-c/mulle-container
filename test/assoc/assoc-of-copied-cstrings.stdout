Get (should be already sorted):
	Den Becher der Fröhlichkeit weihn.=Es haben viel fröhliche Menschen
	Den Ruhenden unter dem Rasen=Es werden viel fröhliche Menschen
	Es haben viel fröhliche Menschen=Den Becher der Fröhlichkeit weihn.
	Es werden viel fröhliche Menschen=Den Ruhenden unter dem Rasen
	Lang nach uns des Lebens sich freun=und gelacht
	Lang vor uns gelebt=Uns Ruhenden unter dem Rasen
	Sei fröhlich=der Becher gebracht.
	Uns Ruhenden unter dem Rasen=Lang vor uns gelebt
	der Becher gebracht.=Sei fröhlich
	und gelacht=Lang nach uns des Lebens sich freun

Enumeration:
	Den Becher der Fröhlichkeit weihn.=Es haben viel fröhliche Menschen
	Den Ruhenden unter dem Rasen=Es werden viel fröhliche Menschen
	Es haben viel fröhliche Menschen=Den Becher der Fröhlichkeit weihn.
	Es werden viel fröhliche Menschen=Den Ruhenden unter dem Rasen
	Lang nach uns des Lebens sich freun=und gelacht
	Lang vor uns gelebt=Uns Ruhenden unter dem Rasen
	Sei fröhlich=der Becher gebracht.
	Uns Ruhenden unter dem Rasen=Lang vor uns gelebt
	der Becher gebracht.=Sei fröhlich
	und gelacht=Lang nach uns des Lebens sich freun

